fix: Simplify / update build instructions. (#713)
aws/aws-encryption-sdk-c: README.md

[Security issue notifications](./CONTRIBUTING.md#security-issue-notifications)

> ## Build recipes with KMS support

Reviewer: Can we be very clear here? "Building the AWS Encryption SDK for C (with support for AWS KMS)"

(The current statement can be interpreted as using KMS to help with the build.)

We will demonstrate some simple build recipes for Linux, Mac, and Windows operating systems. These 

> recipes assume a fresh system with default installs of dependency packages.

Reviewer: Let's define "fresh". Also s/installs/installations. 

I love the voice, really, but many of our customers are global. Let's be clear for them.

> The Linux and Mac recipes install everything in the standard directories in /usr/local. The
> Windows recipe installs everything in an install directory placed at the directory you are in
> when you start the build process. To change the installation directory, if desired, see the Tips
> and Tricks section at the end of this README.

Reviewer: ```suggestion
The Windows instructions install everything in your current directory (where you run the build process). To change the installation directory, see the Tips and Tricks section at the end of this README.
```

The AWS Encryption SDK for C can be used with AWS KMS, but it is not required. If you want to build

> a minimal version of the ESDK without KMS support, see "Minimal C build without KMS support", below.

Reviewer: s/KMS/AWS KMS throughout, please.

For best results when doing a build with KMS integration, do not install aws-c-common directly.
Build and install the AWS SDK for C++, which will build and install aws-c-common for you (see the C++ SDK dependencies
 [here](https://github.com/aws/aws-sdk-cpp/blob/master/third-party/CMakeLists.txt#L18)). If
you install aws-c-common before building the AWS SDK for C++, this will fool the AWS SDK for
C++ install logic, and you will be forced to install several other dependencies manually. Version 1.8.32 of the
AWS SDK for C++ is supported by version v1.0.1 of the AWS Encryption SDK for C.

### If you are working on an EC2 instance, regardless of operating system

> For best results, create an EC2 instance with an instance profile that at a
> minimum has KMS permissions for Encrypt, Decrypt, and GenerateDataKey for
> at least one KMS CMK in your account. You will not need any other AWS

Reviewer: ```suggestion
For best results, create an EC2 instance with an instance profile that at includes AWS KMS permissions for Encrypt, Decrypt, and GenerateDataKey on
at least one KMS CMK in your account. You will not need any other AWS
```

permissions to use the AWS Encryption SDK for C.

### Dependencies

1. OpenSSL 1.0.2 or newer, or 1.1.0 or newer
1. CMake 3.9 or newer
1. C/C++ compiler
1. aws-c-common, typically bundled with the AWS SDK for C++
1. The AWS SDK for C++ version 1.9.35 or newer

> The AWS SDK for C++ and the AWS Encryption SDK for C share dependencies on OpenSSL, aws-c-common, and CMake, and 
> the AWS SDK for C++ has some additional dependencies and prerequisites of its own. See [AWS SDK for
> C++: Getting Started](https://github.com/aws/aws-sdk-cpp#getting-started).

Reviewer: ```suggestion
The AWS SDK for C++ and the AWS Encryption SDK for C share dependencies on OpenSSL, aws-c-common, and CMake. The AWS SDK for C++ has additional dependencies and prerequisites. See [AWS SDK for
C++: Getting Started](https://github.com/aws/aws-sdk-cpp#getting-started).
```

> You need to compile both the AWS Encryption SDK for C and its dependencies as either all

Reviewer: Delete "both." There are more than two things.

shared or all static libraries. These instructions will use all static library builds. Static library
builds are the default for the AWS SDK for C++ and aws-c-common.

> If you would like to build shared libraries, you will need to supply the `-DBUILD_SHARED_LIBS=ON` flag to build

Reviewer: ```suggestion
To build shared libraries, specify the `-DBUILD_SHARED_LIBS=ON` flag to build
```

aws-c-common, the AWS SDK for C++, and the AWS Encryption SDK.

### Linux Build Recipe

> First you will need to build the AWS SDK for C++. That will install the shared dependencies.

Reviewer: ```suggestion
First, build the AWS SDK for C++. That will install the shared dependencies.
```

> If you only need AWS SDK for C++ to use the AWS Encryption SDK, you have the option to build only the AWS KMS SDK.
> Add the `-DBUILD_ONLY="kms"` flag to `cmake` in the instructions provided.

Reviewer: ```suggestion
If you need the AWS SDK for C++ only to use the AWS Encryption SDK, you can build only the AWS KMS SDK.
To do so, add the `-DBUILD_ONLY="kms"` flag to `cmake` as noted in the [AWS SDK for C++ build instructions](https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/setup-linux.html).
```

> [Follow the AWS SDK for C++ build instructions](https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/setup-linux.html).

Reviewer: Delete. Moved this into the previous sentence.

Author: I put the instructions in this order with the intent that I state the required step (first, build the SDK); then I explain what modifications the customer must make to the instructions ("if you only need..."); then I finally link to the directions after clarifying what needs to change.

My concern about putting the instructions link first/combined with the previous sentence is that the customer will correctly click the link and go follow those instructions without realizing that our instructions have a couple of notes that apply to those steps.

Does that make sense? Should I still change this, or change it in some different way?

Reviewer: That's fine.

Now, build and install the AWS Encryption SDK for C:

    git clone --recurse-submodules https://github.com/aws/aws-encryption-sdk-c.git
    mkdir build-aws-encryption-sdk-c && cd build-aws-encryption-sdk-c
    cmake ../aws-encryption-sdk-c
    make && sudo make install ; cd ..

### MacOS Build Recipe

> [Homebrew](https://brew.sh) is a convenient way to obtain build tools on MacOS systems.

Reviewer: s/obtain/get. Also s/on/for (nit).

With Homebrew installed, run the following:

    brew install openssl@1.1 cmake

The AWS SDK for C++ can be installed with Homebrew, which will install the full AWS SDK. If you 

> only need the AWS SDK for C++ to use the AWS Encryption SDK, you have the option to build only the AWS KMS SDK.
> Follow [these directions](https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/setup-linux.html#setup-linux-from-source) 
> and specify the `-DBUILD_ONLY="kms"` flag to `cmake` in the instructions provided.

Reviewer: ```suggestion
need the AWS SDK for C++ only to use the AWS Encryption SDK, you can build only the AWS KMS SDK.
See [these directions](https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/setup-linux.html#setup-linux-from-source)
and specify the `-DBUILD_ONLY="kms"` flag to `cmake` in the instructions provided.
```

Now, build and install the AWS Encryption SDK for C:

    git clone --recurse-submodules https://github.com/aws/aws-encryption-sdk-c.git
    mkdir build-aws-encryption-sdk-c && cd build-aws-encryption-sdk-c
    cmake -G Xcode -DOPENSSL_ROOT_DIR="/usr/local/opt/openssl@1.1" ../aws-encryption-sdk-c 
    xcodebuild -target install; cd ..

### Windows Build Recipe

> **Note**: See the docker-images folder for some Windows build recipes that automate many of these steps.

Reviewer: What docker-images folder? Link? Is there any reason not to use the automation? Should we mark the alternate steps as manual?

Author: https://github.com/aws/aws-encryption-sdk-c/tree/master/docker-images

Requiring customers to use Docker in particular creates another dependency to get started working with the SDK. These generic instructions guide customers through bootstrapping their project generically on any system.

I'll add a link

> Start by installing Visual Studio version 15 or later with the Windows Universal C Runtime and [Git for Windows](https://git-scm.com/download/win).

Reviewer: s/Start by installing/Install

> Use the "x64 Native Tools Command Prompt" for all commands listed here. Run the following commands in the directory where you want to do the build and installation.

Reviewer: Why use that command prompt? Is it the only one that works? Or the only one we tested?

Author: It's the one to have permissions to install the libraries in the default path and to have all of the developer tools populated on the command line without needing to modify or specify the path

Reviewer: Makes sense.

Install Microsoft vcpkg by [following these directions](https://github.com/microsoft/vcpkg#quick-start-windows).

Use vcpkg to integrate with Visual Studio and to install prerequisites:

    mkdir install && mkdir build && cd build
    git clone https://github.com/Microsoft/vcpkg.git
    cd vcpkg && .\bootstrap-vcpkg.bat
    .\vcpkg install curl:x64-windows openssl:x64-windows && cd ..

> You may also want to integrate vcpkg and Visual Studio: `.\vcpkg\vcpkg integrate install`

Reviewer: Does the specified command do the integration? If so:

To integrate vcpkg and Visual Studio (optional), use the following command: `.\vcpkg\vcpkg integrate install`

Author: This is one of the things that I was unhappy with and am fixing, so hold this thought please for the next round

> Build the AWS SDK for C++. This installs the aws-c-common dependency too. If you only need AWS SDK for C++ to use the 

Reviewer: ```suggestion
Build the AWS SDK for C++. This installs the aws-c-common dependency, too. If you need the AWS SDK for C++ only to use the 
```

Place "only" as close to the reference as possible.

> AWS Encryption SDK, you have the option to build only the AWS KMS SDK. Remove `-DBUILD_ONLY=kms` to build the entire AWS SDK for C++.

Reviewer: ```suggestion
AWS Encryption SDK, you can build only the AWS KMS SDK. To build the entire AWS SDK for C++, delete `-DBUILD_ONLY=kms` from ???.
```

To do A, do B. (for left-to-right languages)

    git clone --recurse-submodules https://github.com/aws/aws-sdk-cpp.git
    mkdir build-aws-sdk-cpp && cd build-aws-sdk-cpp
    cmake -DCMAKE_INSTALL_PREFIX="%cd%\..\..\install" -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=ON -DENABLE_UNITY_BUILD=ON -DBUILD_ONLY=kms -DCMAKE_TOOLCHAIN_FILE="%cd%\..\vcpkg\scripts\buildsystems\vcpkg.cmake" -G Ninja ..\aws-sdk-cpp
    cmake --build . && cmake --build . --target install && cd ..

Now, build and install the AWS Encryption SDK for C.

    git clone --recurse-submodules https://github.com/aws/aws-encryption-sdk-c.git
    mkdir build-aws-encryption-sdk-c && cd build-aws-encryption-sdk-c
    cmake -DCMAKE_INSTALL_PREFIX="%cd%\..\..\install" -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=ON -DCMAKE_TOOLCHAIN_FILE"=%cd%\..\vcpkg\scripts\buildsystems\vcpkg.cmake" -G Ninja ..\aws-encryption-sdk-c
    cmake --build . && cmake --build . --target install && cd ..

> ## Building C only, without KMS support

Reviewer: Building the AWS Encryption SDK for C (without support for AWS KMS)

If you don't need AWS KMS support, you can build the AWS Encryption SDK for C without the AWS SDK. This build is C-only with no C++ support
required.

> To build without KMS support, follow the build instructions above for your platform, substituting aws-c-common for aws-sdk-cpp.

Reviewer: s/KMS/AWS KMS (per kenbeer@)
